include/envoy/upstream/cluster_manager.h

```diff
@@ -158,11 +158,23 @@ class ClusterManager {
    *
    * Can return nullptr if there is no host available in the cluster or if the cluster does not
    * exist.
-   */
-  virtual Http::ConnectionPool::Instance* httpConnPoolForCluster(const std::string& cluster,
-                                                                 ResourcePriority priority,
-                                                                 Http::Protocol protocol,
-                                                                 LoadBalancerContext* context) PURE;
+   *
+   * Since the desired protocol might depend on the cluster of the selected host (which could be
+   * different than the top level cluster), we accept a callback that allows protocol selection to
+   * depend on the selected cluster.
+   */
+  virtual Http::ConnectionPool::Instance*
+  httpConnPoolForCluster(const std::string& cluster, ResourcePriority priority,
+                         std::function<Http::Protocol(const ClusterInfo&)> protocol,
```

Comment:
To future proof this why not just pass the host? This would allow potentially for host metadata to be used.